### ISSUE TITLE
Default Parameters + Destructuring  - Pending items

### ISSUE THREAD

**Issue by 0xe** [mozilla/rhino#1641]
Address items pending from https://github.com/mozilla/rhino/pull/1640

- Destructuring with defaults in for loops. This should evaluate to 3, instead throws NPE:

```javascript
function a() {}; (function() { 
            for (let {x = a()} = {}; ; ) { 
                return 3; 
            }
        })()
```

- Reading [Symbol.iterator] from default array/object

```javascript
function f([x = 1, y = 2] = {x: 3, y: 4}) {
 return x + y;
 }
```

This should throw a TypeError as `[Symbol.iterator]` is not implemented for that object on the rhs, instead uses defaults to return `3`.

```javascript
function f([x = 1, y = 2] = {x: 3, y: 4, 
                              [Symbol.iterator]: function*() { 
                                yield 3; 
                                yield 4; 
                              }}) {
                                return x + y;
                              }
```

And this should work because we call into iterator to get the default values. Instead we throw a NPE currently.

**Comment by p-bakker** [comment]
@0xe any chance you'll be picking up this one not that the previous part has been merged?

**Comment by 0xe** [comment]
> @0xe any chance you'll be picking up this one not that the previous part has been merged?

Yep. I'm planning on starting with destructuring with defaults in for loops.

**Comment by 0xe** [comment]
Looking to add support for destructuring with defaults in for loops, I found some problems in how default values were transformed. I've created a PR to fix those first: https://github.com/mozilla/rhino/pull/1708.

**Comment by gbrail** [comment]
Looks like we have made progress since the last time we updated this issue. Can anyone update where we are with this? I'd like to decide if we're going to fix this in the next week or so or move it to a new release.

**Comment by 0xe** [comment]
> Looks like we have made progress since the last time we updated this issue. Can anyone update where we are with this? I'd like to decide if we're going to fix this in the next week or so or move it to a new release.

Sorry, this fell off the radar a bit. I'm still working on fixing this but won't be able to make it by next week.
